### PR TITLE
set CUDA & ROCm stream & handle on Queues

### PR DESCRIPTION
**Pull request by mgates3**
Users have requested the ability to create Queues from a CUDA, ROCm, or SYCL stream/queue or handle, and set the stream/queue or handle. This PR adds those capabilities. Specifically, it adds constructors:
```
Queue( device );          // BLAS++ allocates CUDA/ROCm streams and cuBLAS/rocBLAS handle, or SYCL queue.
Queue( device, stream );  // stream is CUDA/ROCm stream or SYCL queue; BLAS++ allocates cuBLAS/rocBLAS handle.
Queue( device, handle );  // gets stream from cuBLAS/rocBLAS handle.
```
It deprecates the existing constructor, and manages the `batch_chunk` internally:
```
[[deprecated]]
Queue( device, batch_chunk );
```

Additionally, it doesn't allocate streams or workspace memory when the queue is created, but only when needed, typically in a batched BLAS call. This makes queue creation much faster, and eliminates creating unused streams. A standard workspace size is allocated, so the `batch_limit` (renamed `batch_chunk`) is removed from the constructor.

Also, it rewrites the group API for batched gemm so that several groups of pointers can be copied together, which is important if there are many small groups. This also reduces the workspace required.

### CODE REVIEW

**Comment by ayarkhan** [comment]
First attempt here.  I tried to compile on Intel/sunspot and it does not like "streams_ { nullptr }".
There is no conversion from  'std::nullptr_t' to 'stream_t' (aka 'sycl::queue')
```
yarkhan@uan-0002:~/icl/blaspp> make -k
icpx -std=c++17 -O2 -MMD -Wall -Wno-unused-local-typedefs -Wno-unused-function -fopenmp  -I/soft/restricted/CNDA/updates/2022.10.15.001/oneapi/compiler/trunk-20221014/compiler/linux/include -I/soft/restricted/CNDA/updates/2022.10.15.001/oneapi/compiler/trunk-20221014/compiler/linux/include/sycl -fsycl -Wno-deprecated-declarations -fPIC -I./include -c src/device_queue.cc -o src/device_queue.o
src/device_queue.cc:20:16: error: no viable conversion from 'std::nullptr_t' to 'stream_t' (aka 'sycl::queue')
    streams_ { nullptr },  // all streams are null
               ^~~~~~~
```

**Comment by mgates3** [comment]
Timings on guyot (A100 DGX)
Original:
```
    sizeof( Queue ) 240 bytes
    dev 0, create queue avg time 877.880 usec
    dev 0, delete queue avg time 724.963 usec
    fork( 10 )  25.398 usec, join  31.970 usec
    fork( 10 )   6.994 usec, join  29.467 usec
```
I'm not really sure why the 1st fork there is slower than the 2nd fork — in the original code, nothing is allocated in the 1st fork.

Modified:
```
    sizeof( Queue ) 200 bytes
    dev 0, create queue avg time 376.636 usec
    dev 0, delete queue avg time 342.224 usec
    dev 0, create queue avg time 342.875 usec, from stream
    dev 0, delete queue avg time 334.594 usec
    dev 0, create queue avg time   0.273 usec, from handle
    dev 0, delete queue avg time   0.401 usec
    fork( 10 )  21.641 usec, join  31.359 usec
    fork( 10 )   5.861 usec, join  25.229 usec
```
The Queue constructor is significantly faster than before. Creating the cuBLAS handle is the most expensive operation; if that is provided, the Queue constructor takes very little time. Perhaps the BLAS handle should also be allocated on demand, in case the Queue is used only for copying data, no actual BLAS calls.

Here it makes sense that the 1st fork is slower because it creates 9 of 10 streams and all the events.

**Comment by mgates3** [comment]
@ax3l @wavefunction91 This PR implements creating BLAS++ Queues from CUDA/ROCm streams and cuBLAS / rocBLAS handles. The stream or handle can also be set after the Queue is created. I tried to slim down the Queue creation as much as possible — see timings above.

Still working on extending this to oneAPI, but hope to have it finished in the next few days.

Any feedback is welcome.

**Comment by mgates3** [comment]
Updated timings on CUDA (leconte, DGX with 8x V100 GPUs, CUDA 11.4)

```
sh leconte test> ./tester util
BLAS++ version 2023.01.00, id ef7d3f9
[...]
get_device_count = 2    113311.832 usec  # First call is slow!
get_device_count = 2         0.497 usec
get_device_count = 2         0.041 usec
get_device_count = 2         0.040 usec

set_device( 0 )             18.802 usec
set_device( 1 )              0.474 usec
set_device( 0 )              0.224 usec
set_device( 1 )              0.136 usec
set_device( 0 )              0.136 usec
set_device( 1 )              0.138 usec
set_device( 0 )              0.137 usec
set_device( 1 )              0.137 usec

stream_create( 0 )      137609.194 usec  # First call on each device is slow!
stream_create( 0 )           2.573 usec
stream_create( 0 )           1.611 usec
stream_create( 0 )           1.586 usec
stream_create( 1 )      137041.290 usec  # First call on each device is slow!
stream_create( 1 )           2.907 usec
stream_create( 1 )           1.799 usec
stream_create( 1 )           1.931 usec

handle_create( 0 )      452764.913 usec  # First call on each device is slow!
handle_create( 0 )         276.478 usec
handle_create( 0 )         330.626 usec
handle_create( 0 )         322.989 usec
handle_create( 1 )      338639.075 usec  # First call on each device is slow!
handle_create( 1 )         280.268 usec
handle_create( 1 )         268.387 usec
handle_create( 1 )         387.769 usec

handle_set_stream( 0 )      14.895 usec
handle_set_stream( 0 )       0.411 usec
handle_set_stream( 0 )       0.321 usec
handle_set_stream( 0 )       0.322 usec
handle_set_stream( 1 )       0.209 usec
handle_set_stream( 1 )       0.113 usec
handle_set_stream( 1 )       0.120 usec
handle_set_stream( 1 )       0.064 usec

handle_destroy( 0 )        415.988 usec
handle_destroy( 0 )        263.607 usec
handle_destroy( 0 )        257.882 usec
handle_destroy( 0 )        489.465 usec  # Last call on each device is 2x slower.
handle_destroy( 1 )        270.955 usec
handle_destroy( 1 )        258.152 usec
handle_destroy( 1 )        252.303 usec
handle_destroy( 1 )        479.074 usec

stream_destroy( 0 )         24.873 usec
stream_destroy( 0 )          4.059 usec
stream_destroy( 0 )          4.174 usec
stream_destroy( 0 )          4.061 usec
stream_destroy( 1 )          5.383 usec
stream_destroy( 1 )          3.892 usec
stream_destroy( 1 )          3.736 usec
stream_destroy( 1 )          3.790 usec

# Since cuBLAS handles have been created above on each device,
# these are fast. The first one on each device is still 2x slower.
# I guess since all previous cuBLAS handles have been destroyed,
# it has to do some re-init of cuBLAS; see delete queues below.
test_queue
sizeof( Queue ) 200 bytes
blas::Queue( 0 )           575.574 usec  # First call on each device is 2x slower.
blas::Queue( 0 )           281.908 usec
blas::Queue( 0 )           300.618 usec
blas::Queue( 0 )           299.418 usec
blas::Queue( 1 )           565.642 usec  # First call on each device is 2x slower.
blas::Queue( 1 )           335.555 usec
blas::Queue( 1 )           270.231 usec
blas::Queue( 1 )           312.024 usec

# Interestingly, the last one is slower. I guess when
# the last cuBLAS handle is destroyed on each device,
# it deallocates some extra stuff.
delete queues[ 0 ]         308.421 usec
delete queues[ 0 ]         262.336 usec
delete queues[ 0 ]         254.157 usec
delete queues[ 0 ]         545.166 usec  # Last one on each device is 2x slower.
delete queues[ 1 ]         257.127 usec
delete queues[ 1 ]         249.743 usec
delete queues[ 1 ]         382.559 usec
delete queues[ 1 ]         528.352 usec  # Last one on each device is 2x slower.

# Creating a Queue from a stream shows similar
# times to creating from scratch above.
test_queue_from_stream
blas::Queue( 0, stream )      280.839 usec, from stream
blas::Queue( 0, stream )      274.816 usec, from stream
blas::Queue( 0, stream )      268.785 usec, from stream
blas::Queue( 0, stream )      274.086 usec, from stream
blas::Queue( 1, stream )      292.027 usec, from stream
blas::Queue( 1, stream )      277.439 usec, from stream
blas::Queue( 1, stream )      275.268 usec, from stream
blas::Queue( 1, stream )      280.729 usec, from stream

delete queues[ 0 ]            266.165 usec
delete queues[ 0 ]            249.756 usec
delete queues[ 0 ]            260.650 usec
delete queues[ 0 ]            267.742 usec
delete queues[ 1 ]            256.752 usec
delete queues[ 1 ]            246.084 usec
delete queues[ 1 ]            257.401 usec
delete queues[ 1 ]            257.265 usec

# Creating a Queue from a cuBLAS handle is fast.
blas::Queue( 0, handle )        8.431 usec, from handle
blas::Queue( 0, handle )        1.038 usec, from handle
blas::Queue( 0, handle )        0.208 usec, from handle
blas::Queue( 0, handle )        0.177 usec, from handle
blas::Queue( 1, handle )        0.267 usec, from handle
blas::Queue( 1, handle )        0.199 usec, from handle
blas::Queue( 1, handle )        0.162 usec, from handle
blas::Queue( 1, handle )        0.162 usec, from handle

# Deleting a Queue is fast, since it doesn't delete
# the user-provided cuBLAS handle.
delete queues[ 0 ]              0.992 usec
delete queues[ 0 ]              0.356 usec
delete queues[ 0 ]              0.359 usec
delete queues[ 0 ]              0.345 usec
delete queues[ 1 ]              0.437 usec
delete queues[ 1 ]              0.348 usec
delete queues[ 1 ]              0.351 usec
delete queues[ 1 ]              0.362 usec

# Here the 1st fork is allocating necessary streams and events.
# After the 1st fork on a queue, subsequent ones should be faster.
# The more streams we fork, the more expensive it gets.
test_queue_fork
queue( dev 0 )
fork( 1 )      24.712 usec, join       1.355 usec (calls blas::scal)
fork( 1 )       3.183 usec, join       0.091 usec (no call)
fork( 1 )       1.517 usec, join       0.074 usec (calls blas::scal)
fork( 1 )       1.797 usec, join       0.058 usec (no call)
fork( 1 )       5.475 usec, join       0.054 usec (calls blas::scal)
fork( 1 )       1.557 usec, join       0.077 usec (no call)
After first one, fork should be faster.
    ~queue
queue( dev 0 )
fork( 3 )      19.031 usec, join       4.292 usec (calls blas::scal)
fork( 3 )       2.528 usec, join       4.302 usec (no call)
fork( 3 )       2.437 usec, join       4.033 usec (calls blas::scal)
fork( 3 )       2.126 usec, join       3.508 usec (no call)
fork( 3 )       2.506 usec, join       3.626 usec (calls blas::scal)
fork( 3 )       2.083 usec, join       3.344 usec (no call)
After first one, fork should be faster.
    ~queue
queue( dev 0 )
fork( 10 )     233.519 usec, join      18.022 usec (calls blas::scal)
fork( 10 )       4.465 usec, join      17.956 usec (no call)
fork( 10 )       4.834 usec, join      14.983 usec (calls blas::scal)
fork( 10 )       4.146 usec, join      16.498 usec (no call)
fork( 10 )       4.304 usec, join      18.426 usec (calls blas::scal)
fork( 10 )       3.930 usec, join      16.693 usec (no call)
After first one, fork should be faster.
    ~queue
```

**Comment by mgates3** [comment]
Updated timings for ROCm (dopamine, 2x MI210, ROCm 5.4.2)

```
sh dopamine test> ./tester util
BLAS++ version 2023.01.00, id ef7d3f9
[...]
get_device_count = 2     71495.264 usec  # First call slow, though not as slow as CUDA
get_device_count = 2         0.310 usec
get_device_count = 2         0.140 usec
get_device_count = 2         0.070 usec

set_device( 0 )              3.540 usec
set_device( 1 )              0.160 usec
set_device( 0 )              0.080 usec
set_device( 1 )              0.070 usec
set_device( 0 )              0.070 usec
set_device( 1 )              0.070 usec
set_device( 0 )              0.080 usec
set_device( 1 )              0.070 usec

stream_create( 0 )      220897.917 usec  # First call on each device slow
stream_create( 0 )        7972.480 usec  # Rest are faster, but much slower than CUDA (~ 2)
stream_create( 0 )        6042.327 usec
stream_create( 0 )        6523.723 usec
stream_create( 1 )      185954.102 usec  # First call on each device slow
stream_create( 1 )        7718.693 usec
stream_create( 1 )        6997.529 usec
stream_create( 1 )        6869.250 usec

handle_create( 0 )         133.969 usec  # First call slow
handle_create( 0 )          16.190 usec  # Much faster than CUDA (~ 300)
handle_create( 0 )          15.560 usec
handle_create( 0 )          12.670 usec
handle_create( 1 )          15.310 usec
handle_create( 1 )          11.700 usec
handle_create( 1 )          12.710 usec
handle_create( 1 )          11.689 usec

handle_set_stream( 0 )       7.449 usec
handle_set_stream( 0 )       0.580 usec
handle_set_stream( 0 )       0.370 usec
handle_set_stream( 0 )       0.460 usec
handle_set_stream( 1 )       0.420 usec
handle_set_stream( 1 )       0.360 usec
handle_set_stream( 1 )       0.260 usec
handle_set_stream( 1 )       0.360 usec

handle_destroy( 0 )         92.649 usec
handle_destroy( 0 )         28.000 usec  # In most cases, much faster than CUDA (~ 300)
handle_destroy( 0 )         29.680 usec
handle_destroy( 0 )         25.489 usec
handle_destroy( 1 )        235.758 usec
handle_destroy( 1 )         26.610 usec
handle_destroy( 1 )         30.270 usec
handle_destroy( 1 )         26.249 usec

stream_destroy( 0 )        854.113 usec  # Much slower than CUDA (~ 4)
stream_destroy( 0 )        689.834 usec
stream_destroy( 0 )        696.454 usec
stream_destroy( 0 )        685.314 usec
stream_destroy( 1 )        649.545 usec
stream_destroy( 1 )        546.825 usec
stream_destroy( 1 )        572.405 usec
stream_destroy( 1 )        588.355 usec

test_queue
sizeof( Queue ) 200 bytes
blas::Queue( 0 )          5888.808 usec
blas::Queue( 0 )          5660.050 usec
blas::Queue( 0 )         21262.573 usec  # Odd that time goes up
blas::Queue( 0 )         20803.778 usec
blas::Queue( 1 )          5990.918 usec
blas::Queue( 1 )          5959.757 usec
blas::Queue( 1 )          6139.337 usec
blas::Queue( 1 )          6449.224 usec

delete queues[ 0 ]         686.194 usec
delete queues[ 0 ]         607.645 usec
delete queues[ 0 ]         583.275 usec
delete queues[ 0 ]         607.994 usec
delete queues[ 1 ]         590.965 usec
delete queues[ 1 ]         593.175 usec
delete queues[ 1 ]         600.174 usec
delete queues[ 1 ]         602.055 usec

# As you would expect from above timings,
# creating Queue from ROCm stream is fast
test_queue_from_stream
blas::Queue( 0, stream )       18.710 usec, from stream
blas::Queue( 0, stream )       14.810 usec, from stream
blas::Queue( 0, stream )       58.889 usec, from stream
blas::Queue( 0, stream )       39.630 usec, from stream
blas::Queue( 1, stream )       42.579 usec, from stream
blas::Queue( 1, stream )       37.390 usec, from stream
blas::Queue( 1, stream )       35.670 usec, from stream
blas::Queue( 1, stream )       38.889 usec, from stream

delete queues[ 0 ]             44.660 usec
delete queues[ 0 ]             31.719 usec
delete queues[ 0 ]             29.529 usec
delete queues[ 0 ]             26.230 usec
delete queues[ 1 ]             31.670 usec
delete queues[ 1 ]             27.449 usec
delete queues[ 1 ]             26.860 usec
delete queues[ 1 ]             28.150 usec

# And creating it from a rocBLAS handle is even faster.
blas::Queue( 0, handle )        4.770 usec, from handle
blas::Queue( 0, handle )        0.150 usec, from handle
blas::Queue( 0, handle )        0.150 usec, from handle
blas::Queue( 0, handle )        0.130 usec, from handle
blas::Queue( 1, handle )        0.390 usec, from handle
blas::Queue( 1, handle )        0.100 usec, from handle
blas::Queue( 1, handle )        0.470 usec, from handle
blas::Queue( 1, handle )        0.260 usec, from handle

delete queues[ 0 ]              0.409 usec
delete queues[ 0 ]              0.200 usec
delete queues[ 0 ]              0.360 usec
delete queues[ 0 ]              0.200 usec
delete queues[ 1 ]              0.199 usec
delete queues[ 1 ]              0.190 usec
delete queues[ 1 ]              0.200 usec
delete queues[ 1 ]              0.310 usec

# Forking is fairly expensive compared to CUDA, since creating streams is expensive.
test_queue_fork
queue( dev 0 )
fork( 1 )      13.270 usec, join       1.840 usec (calls blas::scal)
fork( 1 )       6.770 usec, join       0.050 usec (no call)
fork( 1 )       3.070 usec, join       0.030 usec (calls blas::scal)
fork( 1 )       3.640 usec, join       0.030 usec (no call)
fork( 1 )       3.090 usec, join       0.030 usec (calls blas::scal)
fork( 1 )       2.510 usec, join       0.030 usec (no call)
After first one, fork should be faster.
    ~queue
queue( dev 0 )
fork( 3 )   22520.253 usec, join       6.910 usec (calls blas::scal)
fork( 3 )       5.720 usec, join       7.720 usec (no call)
fork( 3 )       9.720 usec, join      68.079 usec (calls blas::scal)
fork( 3 )      13.560 usec, join       7.870 usec (no call)
fork( 3 )       6.000 usec, join      11.150 usec (calls blas::scal)
fork( 3 )      12.780 usec, join      15.110 usec (no call)
After first one, fork should be faster.
    ~queue
queue( dev 0 )
fork( 10 )   37687.691 usec, join      67.840 usec (calls blas::scal)
fork( 10 )      30.969 usec, join      37.689 usec (no call)
fork( 10 )      22.670 usec, join      51.870 usec (calls blas::scal)
fork( 10 )      25.490 usec, join      46.919 usec (no call)
fork( 10 )      36.580 usec, join      58.210 usec (calls blas::scal)
fork( 10 )      42.689 usec, join      69.310 usec (no call)
After first one, fork should be faster.
    ~queue
```

File: src/device_batch_gemm.cc [hunk 1]
```diff
@@ -82,13 +82,16 @@ void gemm(
         device_blas_int ldb_ = to_device_blas_int( ldb[0] );
         device_blas_int ldc_ = to_device_blas_int( ldc[0] );
 
-        size_t batch_limit = queue.get_batch_limit();
-        scalar_t** dAarray = (scalar_t**) queue.get_dev_ptr_array( );
-        scalar_t** dBarray = dAarray + batch_limit;
-        scalar_t** dCarray = dBarray + batch_limit;
+        // gemm needs 3 arrays (A, B, and C).
+        size_t max_chunk = MaxBatchChunk;
+        queue.work_resize<void*>( 3*max_chunk );
```
Comment (cayrols) [inline src/device_batch_gemm.cc:87]: Should not we resize only when we do not have enough space?

File: src/device_batch_trsm.cc [hunk 2]
```diff
@@ -84,12 +84,17 @@ void trsm(
...
         }
 
-        size_t batch_limit = queue.get_batch_limit();
-        scalar_t** dAarray = (scalar_t**) queue.get_dev_ptr_array( );
-        scalar_t** dBarray = dAarray + batch_limit;
+        // trsm needs only 2 ptr arrays (A and B). Allocate usual
+        // 3*max_chunk used by gemm, but then split in 2 instead of 3.
+        size_t max_chunk = MaxBatchChunk;
+        queue.work_resize<void*>( 3*max_chunk );
+        max_chunk = 3*max_chunk / 2;
```
Comment (cayrols) [inline src/device_batch_trsm.cc:91]: Why give 3/2 of max_chunk to a matrix when, from gemm point of view, one max_chunk is used?

File: src/device_queue.cc [hunk 3]
```diff
@@ -4,282 +4,218 @@
...
+    /// Change the CUDA or HIP stream used in the BLAS++ queue.
+    /// Kernels executing on the current stream will continue.
+    /// Throws an error if in fork mode.
+    void Queue::set_stream( stream_t& stream )
+    {
+        if (current_stream_index_ != 0)
+            throw blas::Error( "can't set stream in fork mode", __func__ );
+
+        if (own_default_stream_) {
+            stream_destroy( streams_[ 0 ] );
```
Comment (cayrols) [inline src/device_queue.cc:130]: Do not we need to check if a handle is associated to the stream and destroy it if so?

File: src/device_queue.cc [hunk 4]
```diff
@@ -4,282 +4,218 @@
...
+    /// Change the cuBLAS or rocBLAS handle used in the BLAS++ queue.
+    /// Gets the stream from the handle.
+    /// Throws an error if in fork mode.
+    void Queue::set_handle( handle_t& handle )
+    {
+        if (current_stream_index_ != 0)
+            throw blas::Error( "can't set stream in fork mode", __func__ );
+
+        if (own_default_stream_) {
+            stream_destroy( streams_[ 0 ] );
```
Comment (cayrols) [inline src/device_queue.cc:147]: Same question here.

File: src/device_queue.cc [hunk 5]
```diff
@@ -4,282 +4,218 @@
...
+                    event_destroy( events_[ i ] );
+                    events_[ i ] = nullptr;
+                }
             }
 
-        #elif defined(BLAS_HAVE_ONEMKL)
-            device_free( work_, *this );
-            delete default_stream_;
+        #elif defined( BLAS_HAVE_ONEMKL )
+            //delete streams_[ 0 ];
```
Comment (cayrols) [inline src/device_queue.cc:218]: Is it a TODO?

File: test/test_util.cc [hunk 6]
```diff
@@ -346,14 +349,486 @@ void test_make_scalar()
...
+        blas::Queue queue( dev );
+
+        float* x = blas::device_malloc<float>( n, queue );
+
+        auto orig_stream = queue.stream();
+        #if defined( BLAS_HAVE_CUBLAS ) || defined( BLAS_HAVE_ROCBLAS )
+            auto handle = queue.handle();
+        #endif
+
+        std::vector< blas::Queue::stream_t > streams( num_streams );  //, nullptr );
```
Comment (cayrols) [inline test/test_util.cc:761]: Remove comment.

**Review by cayrols — APPROVED** [review]
Looks good to me.

File: src/device_queue.cc (hunk 3 above)
Comment (mgates3) [reply src/device_queue.cc:130]: I don't think so. There's only one handle, and we change that handle to use this stream below.

File: src/device_queue.cc (hunk 5 above)
Comment (mgates3) [reply src/device_queue.cc:218]: Fixed. It's implicitly deleted.

File: src/device_batch_gemm.cc (hunk 1 above)
Comment (mgates3) [reply src/device_batch_gemm.cc]: Per discussion, changed to `work_ensure_size`.

**Pull request merged**